db/db_bolt.go

```diff
@@ -305,13 +315,32 @@ func (b *BoltDB) WriteBatch(kvsb batch.KVStoreBatch) (err error) {
 	kvsb.Lock()
 	defer kvsb.Unlock()
 
+	type doubleKey struct {
+		ns  string
+		key string
+	}
+	// remove duplicate keys, only keep the last write for each key
+	entryMap := make(map[doubleKey]*batch.WriteInfo)
+	for i := kvsb.Size() - 1; i >= 0; i-- {
+		write, e := kvsb.Entry(i)
+		if e != nil {
+			return e
+		}
+		// only handle Put and Delete
+		if write.WriteType() != batch.Put && write.WriteType() != batch.Delete {
+			continue
+		}
+		k := doubleKey{ns: write.Namespace(), key: string(write.Key())}
+		if _, ok := entryMap[k]; !ok {
+			entryMap[k] = write
+		}
+	}
+	boltdbMtc.WithLabelValues(b.path, "entrySize").Set(float64(kvsb.Size()))
+	boltdbMtc.WithLabelValues(b.path, "uniqueEntrySize").Set(float64(len(entryMap)))
 	for c := uint8(0); c < b.config.NumRetries; c++ {
 		if err = b.db.Update(func(tx *bolt.Tx) error {
-			for i := 0; i < kvsb.Size(); i++ {
-				write, e := kvsb.Entry(i)
-				if e != nil {
-					return e
-				}
+			// though range entryMap is random, but it doesn't matter which key is processed first
```

Comment:
will the random sequence cause the nondeterministic md5 value of Snapshot file? For example, will the binary file(*.db) be same between `bucket.Put(A,{}) -> bucket.Put(B,{})` and `bucket.Put(B,{}) -> bucket.Put(A,{})`?